sourced/ml/algorithms/token_parser.py

```diff
@@ -1,29 +1,50 @@
+from enum import Enum
 import re
 
 import Stemmer
 
 
+class TokenMeta(Enum):
+    """Metadata that should allow to reconstruct initial identifier from a list of tokens."""
+    DELIMITER = 1
+    TOKEN_UPPER = 2
+    TOKEN_LOWER = 3
+    TOKEN_STARTS_WITH_UPPER = 4
+
+
 class TokenParser:
     """
     Common utilities for splitting and stemming tokens.
     """
-    NAME_BREAKUP_RE = re.compile(r"[^a-zA-Z]+")  #: Regexp to split source code identifiers.
-    STEM_THRESHOLD = 6  #: We do not stem split parts shorter than or equal to this size.
-    MAX_TOKEN_LENGTH = 256  #: We cut identifiers longer than this value.
-    MIN_SPLIT_LENGTH = 3  #: We do not split source code identifiers shorter than this value.
-    DEFAULT_SINGLE_SHOT = False  #: True if we do not want to join small identifiers to next one.
+    # Regexp to split source code identifiers
+    NAME_BREAKUP_RE = re.compile(r"[^a-zA-Z]+")
+    NAME_BREAKUP_KEEP_DELIMITERS_RE = re.compile(r"([^a-zA-Z]+)")  # ... and keep delimiters
```

Comment:
This is not required at all. Everything which is between groups detected by `NAME_BREAKUP_RE` should be taken.